node/testing/src/scenarios/driver.rs

```diff
@@ -0,0 +1,363 @@
+use std::time::{Duration, Instant};
+
+use libp2p::Multiaddr;
+use node::{
+    event_source::Event,
+    p2p::{
+        connection::outgoing::P2pConnectionOutgoingInitOpts, P2pConnectionEvent, P2pEvent,
+        P2pListenEvent, P2pListenerId, PeerId,
+    },
+    State,
+};
+
+use crate::{cluster::ClusterNodeId, node::RustNodeTestingConfig, scenario::ScenarioStep};
+
+use super::ClusterRunner;
+
+pub fn match_addr_with_port_and_peer_id(
+    port: u16,
+    peer_id: PeerId,
+) -> impl Fn(&P2pConnectionOutgoingInitOpts) -> bool {
+    move |conn_opt| match conn_opt {
+        P2pConnectionOutgoingInitOpts::LibP2P(libp2p_opts) => {
+            &libp2p_opts.peer_id == &peer_id && libp2p_opts.port == port
+        }
+        _ => false,
+    }
+}
+
+pub fn get_peers_iter(
+    data: &serde_json::Value,
+) -> Option<impl Iterator<Item = Option<(&str, i64, &str)>>> {
+    let iter = data
+        .as_object()?
+        .get("data")?
+        .get("getPeers")?
+        .as_array()?
+        .iter()
+        .map(|elt| {
+            let elt = elt.as_object()?;
+            let host = elt.get("host")?.as_str()?;
+            let port = elt.get("libp2pPort")?.as_i64()?;
+            let peer_id = elt.get("peerId")?.as_str()?;
+            Some((host, port, peer_id))
+        });
+    Some(iter)
+}
+
+pub const PEERS_QUERY: &str = r#"query {
+  getPeers {
+    host
+    libp2pPort
+    peerId
+  }
+}"#;
+
+pub fn connection_finalized_event(
+    pred: impl Fn(ClusterNodeId, &PeerId) -> bool,
+) -> impl Fn(ClusterNodeId, &Event, &State) -> bool {
+    move |node_id, event, _| {
+        matches!(
+            event,
+            Event::P2p(P2pEvent::Connection(P2pConnectionEvent::Finalized(peer, res))) if pred(node_id, peer) && res.is_ok()
+        )
+    }
+}
+
+pub fn connection_finalized_with_res_event(
+    pred: impl Fn(ClusterNodeId, &PeerId, &Result<(), String>) -> bool,
+) -> impl Fn(ClusterNodeId, &Event, &State) -> bool {
+    move |node_id, event, _| {
+        matches!(
+            event,
+            Event::P2p(P2pEvent::Connection(P2pConnectionEvent::Finalized(peer, res))) if pred(node_id, peer, res)
+        )
+    }
+}
+
+pub fn as_listen_new_addr_event(event: &Event) -> Option<(&Multiaddr, &P2pListenerId)> {
+    if let Event::P2p(P2pEvent::Listen(P2pListenEvent::NewListenAddr { listener_id, addr })) = event
+    {
+        Some((addr, listener_id))
+    } else {
+        None
+    }
+}
+
+pub fn as_connection_finalized_event(event: &Event) -> Option<(&PeerId, &Result<(), String>)> {
+    if let Event::P2p(P2pEvent::Connection(P2pConnectionEvent::Finalized(peer, res))) = event {
+        Some((peer, res))
+    } else {
+        None
+    }
+}
+
+pub fn identify_event(peer_id: PeerId) -> impl Fn(ClusterNodeId, &Event, &State) -> bool {
+    move |_, event, _| {
+        matches!(
+            event,
+            Event::P2p(P2pEvent::Libp2pIdentify(peer, _)) if peer == &peer_id
+        )
+    }
+}
+
+pub struct Driver<'cluster> {
```

Comment:
Do we need more abstraction on top of `ClusterRunner`? Roles between `Driver` and `ClusterRunner` seems to be the same. Could we move this logic to `ClusterRunner` instead? If it's the name, we can rename it to `Driver` if that's better.

Reply:
Well, driver is a bunch of utilities on top of the cluster runner. Until it is more clear what would be separation, let's keep it like this.